File: packages/core/src/merkle.js
import { sha256, LEAF_PREFIX, NODE_PREFIX, equalBytes } from './hash.js';

/**
 * RFC 6962 Merkle tree — the same construction Certificate Transparency uses.
 *
 * Two proofs matter here, and they answer different questions:
 *
 *   inclusion   "receipt #4,201 is in the log you published" — for anyone
 *               holding a single receipt who wants to tie it to a root.
 *   consistency "the log you published today still contains, unmodified,
 *               everything you published last Tuesday" — this is the one that
 *               makes deletion and back-dating detectable. Without it an
 *               operator can quietly rewrite history and re-sign a new root.
 */

/** Root of the empty tree, per RFC 6962 §2.1. */
const EMPTY_ROOT = sha256(Buffer.alloc(0));

/**
 * @param {Buffer} data  The canonical bytes of a receipt.
 * @returns {Buffer}
 */
export function leafHash(data) {
  return sha256(LEAF_PREFIX, data);
}

/**
 * @param {Buffer} left
 * @param {Buffer} right
 * @returns {Buffer}
 */
export function nodeHash(left, right) {
  return sha256(NODE_PREFIX, left, right);
}

/**
 * Largest power of two strictly less than n. Defined only for n > 1.
 *
 * @param {number} n
 * @returns {number}
 */
function splitPoint(n) {
  let k = 1;
  while (k << 1 < n) k <<= 1;
  return k;
}

/**
 * Merkle tree head over a slice of leaf hashes.
 *
 * @param {Buffer[]} leaves
 * @returns {Buffer}
 */
export function merkleRoot(leaves) {
  if (leaves.length === 0) return EMPTY_ROOT;
  if (leaves.length === 1) return leaves[0];
  const k = splitPoint(leaves.length);
  return nodeHash(merkleRoot(leaves.slice(0, k)), merkleRoot(leaves.slice(k)));
}

/**
 * An append-only tree that keeps the root cheap to maintain.
 *
 * Internally this is a binary counter of complete subtrees: appending a leaf
 * pushes a size-1 subtree and merges equal-sized neighbours, so an append is
 * O(log n) rather than the O(n) a full recomputation would cost. The root is
 * the right-to-left fold of that stack, which is exactly RFC 6962's
 * "split at the largest power of two" decomposition read from the other end.
 */
export class MerkleTree {
  /** @param {Buffer[]} [leaves] Existing leaf hashes, oldest first. */
  constructor(leaves = []) {
    /** @type {Buffer[]} */
    this.leaves = [];
    /** @type {{ size: number, hash: Buffer }[]} */
    this._stack = [];
    /**
     * Cached level structure, built on demand for proof generation.
     * `_levels[0]` is the leaves; each subsequent level pairs adjacent nodes
     * and promotes an unpaired last node. Invalidated on append.
     * @type {Buffer[][]|null}
     */
    this._levels = null;
    for (const l of leaves) this.append(l);
  }

  /**
   * Build the level structure, so an inclusion proof is a walk up the tree
   * rather than a recomputation of it.
   *
   * The naive recursive PATH is O(n) per proof because it rebuilds every
   * subtree root it passes. That is invisible on a toy log and quadratic on a
   * real one: generating proofs for all n entries of an export costs O(n²),
   * which measured at 41 seconds for 4,000 receipts and would be hours for a
   * hundred thousand. One O(n) build amortised across every proof turns each
   * one into O(log n).
   *
   * Pairing adjacent nodes and promoting the odd one out is exactly RFC 6962's
   * "split at the largest power of two" decomposition, read bottom-up.
   *
   * @returns {Buffer[][]}
   */
  _buildLevels() {
    if (this._levels) return this._levels;

    /** @type {Buffer[][]} */
    const levels = [this.leaves.slice()];
    while (levels[levels.length - 1].length > 1) {
      const below = levels[levels.length - 1];
      /** @type {Buffer[]} */
      const above = [];
      for (let i = 0; i < below.length; i += 2) {
        // An unpaired final node rises unchanged; it pairs at a higher level.
        above.push(i + 1 < below.length ? nodeHash(below[i], below[i + 1]) : below[i]);
      }
      levels.push(above);
    }

    this._levels = levels;
    return levels;
  }

  /** @returns {number} */
  get size() {
    return this.leaves.length;
  }

  /**
   * @param {Buffer} hash  A leaf hash, already domain-tagged by `leafHash`.
   * @returns {number} The index the leaf was stored at.
   */
  append(hash) {
    this.leaves.push(hash);
    this._levels = null;
    this._stack.push({ size: 1, hash });
    while (this._stack.length > 1) {
      const right = this._stack[this._stack.length - 1];
      const left = this._stack[this._stack.length - 2];
      if (left.size !== right.size) break;
      this._stack.splice(-2, 2, {
        size: left.size + right.size,
        hash: nodeHash(left.hash, right.hash),
      });
    }
    return this.leaves.length - 1;
  }

  /** @returns {Buffer} */
  get root() {
    if (this._stack.length === 0) return EMPTY_ROOT;
    let acc = this._stack[this._stack.length - 1].hash;
    for (let i = this._stack.length - 2; i >= 0; i--) {
      acc = nodeHash(this._stack[i].hash, acc);
    }
    return acc;
  }

  /**
   * Root the tree had when it held exactly `size` leaves. Needed to check a
   * historical checkpoint against the log as it stands now.
   *
   * @param {number} size
   * @returns {Buffer}
   */
  rootAt(size) {
    if (size < 0 || size > this.leaves.length) {
      throw new RangeError(`size ${size} out of range for tree of ${this.leaves.length}`);
    }
    return merkleRoot(this.leaves.slice(0, size));
  }

  /**
   * @param {number} index
   * @param {number} [treeSize]  Defaults to the current size.
   * @returns {Buffer[]}
   */
  inclusionProof(index, treeSize = this.leaves.length) {
    // The fast path covers the overwhelmingly common case: a proof against the
    // tree as it stands now. A proof against a historical size still needs the
    // recursive form, which is fine — that is a rare, one-off request.
    if (treeSize !== this.leaves.length) {
      return inclusionProof(this.leaves.slice(0, treeSize), index);
    }
    if (!Number.isInteger(index) || index < 0 || index >= this.leaves.length) {
      throw new RangeError(`leaf index ${index} out of range for ${this.leaves.length} leaves`);
    }

    const levels = this._buildLevels();
    /** @type {Buffer[]} */
    const proof = [];
    let i = index;

    for (let depth = 0; depth < levels.length - 1; depth++) {
      const level = levels[depth];
      const sibling = i ^ 1;
      // No sibling means this node was promoted unchanged; nothing to prove
      // at this level, and the index simply rises.
      if (sibling < level.length) proof.push(level[sibling]);
      i >>>= 1;
    }
    return proof;
  }

  /**
   * @param {number} fromSize
   * @param {number} [toSize]  Defaults to the current size.
   * @returns {Buffer[]}
   */
  consistencyProof(fromSize, toSize = this.leaves.length) {
    return consistencyProof(this.leaves.slice(0, toSize), fromSize);
  }
}

/**
 * PATH(m, D[n]) from RFC 6962 §2.1.1 — the sibling hashes from a leaf up to
 * the root, bottom-first.
 *
 * @param {Buffer[]} leaves
 * @param {number} index
 * @returns {Buffer[]}
 */
export function inclusionProof(leaves, index) {
  if (!Number.isInteger(index) || index < 0 || index >= leaves.length) {
    throw new RangeError(`leaf index ${index} out of range for ${leaves.length} leaves`);
  }
  if (leaves.length === 1) return [];
  const k = splitPoint(leaves.length);
  if (index < k) {
    return [...inclusionProof(leaves.slice(0, k), index), merkleRoot(leaves.slice(k))];
  }
  return [...inclusionProof(leaves.slice(k), index - k), merkleRoot(leaves.slice(0, k))];
}

/**
 * Recompute the root a proof implies, then compare. Returns false rather than
 * throwing on malformed input: a verifier is fed hostile data by definition.
 *
 * @param {object} args
 * @param {Buffer} args.leafHash
 * @param {number} args.index
 * @param {number} args.treeSize
 * @param {Buffer[]} args.proof
 * @param {Buffer} args.root
 * @returns {boolean}
 */
export function verifyInclusion({ leafHash: leaf, index, treeSize, proof, root }) {
  if (!Number.isInteger(index) || !Number.isInteger(treeSize)) return false;
  if (index < 0 || treeSize <= 0 || index >= treeSize) return false;

  let fn = index;
  let sn = treeSize - 1;
  let acc = leaf;

  for (const sibling of proof) {
    if (sn === 0) return false;
    if (sibling.length !== 32) return false;
    if ((fn & 1) === 1 || fn === sn) {
      acc = nodeHash(sibling, acc);
      while (fn !== 0 && (fn & 1) === 0) {
        fn >>>= 1;
        sn >>>= 1;
      }
    } else {
      acc = nodeHash(acc, sibling);
    }
    fn >>>= 1;
    sn >>>= 1;
  }

  // A proof that runs out before reaching the root proves nothing.
  if (sn !== 0) return false;
  return equalBytes(acc, root);
}

/**
 * PROOF(m, D[n]) from RFC 6962 §2.1.2 — evidence that the size-m tree is a
 * prefix of the size-n tree.
 *
 * @param {Buffer[]} leaves  The larger (current) leaf set.
 * @param {number} fromSize  The smaller, historical size m.
 * @returns {Buffer[]}
 */
export function consistencyProof(leaves, fromSize) {
  const n = leaves.length;
  if (!Number.isInteger(fromSize) || fromSize < 0 || fromSize > n) {
    throw new RangeError(`fromSize ${fromSize} out of range for ${n} leaves`);
  }
  if (fromSize === 0 || fromSize === n) return [];
  return subProof(leaves, fromSize, true);
}

/**
 * @param {Buffer[]} leaves
 * @param {number} m
 * @param {boolean} isRootOfOld  True while the old root is still an exact
 *   subtree root of the path we are walking, so it need not be sent.
 * @returns {Buffer[]}
 */
function subProof(leaves, m, isRootOfOld) {
  const n = leaves.length;
  if (m === n) return isRootOfOld ? [] : [merkleRoot(leaves)];
  const k = splitPoint(n);
  if (m <= k) {
    return [...subProof(leaves.slice(0, k), m, isRootOfOld), merkleRoot(leaves.slice(k))];
  }
  return [...subProof(leaves.slice(k), m - k, false), merkleRoot(leaves.slice(0, k))];
}

/**
 * Check that `secondRoot` is an append-only extension of `firstRoot`.
 *
 * This is the check an auditor runs on a schedule. If it ever fails, the log
 * operator has rewritten history — no further interpretation required.
 *
 * @param {object} args
 * @param {number} args.firstSize
 * @param {number} args.secondSize
 * @param {Buffer} args.firstRoot
 * @param {Buffer} args.secondRoot
 * @param {Buffer[]} args.proof
 * @returns {boolean}
 */
export function verifyConsistency({ firstSize, secondSize, firstRoot, secondRoot, proof }) {
  if (!Number.isInteger(firstSize) || !Number.isInteger(secondSize)) return false;
  if (firstSize < 0 || secondSize < firstSize) return false;

  if (firstSize === secondSize) {
    return proof.length === 0 && equalBytes(firstRoot, secondRoot);
  }
  // Everything extends the empty tree; there is nothing to prove.
  if (firstSize === 0) return proof.length === 0;

  let fn = firstSize - 1;
  let sn = secondSize - 1;
  // Strip the trailing ones: those levels are shared by both roots.
  while ((fn & 1) === 1) {
    fn >>>= 1;
    sn >>>= 1;
  }

  if (proof.length === 0) return false;
  let i = 0;
  let fr;
  let sr;
  if (fn !== 0) {
    // The old root was not a complete subtree, so its left edge is supplied.
    fr = proof[i];
    sr = proof[i];
    i++;
  } else {
    fr = firstRoot;
    sr = firstRoot;
  }

  for (; i < proof.length; i++) {
    const p = proof[i];
    if (sn === 0) return false;
    if (p.length !== 32) return false;
    if ((fn & 1) === 1 || fn === sn) {
      fr = nodeHash(p, fr);
      sr = nodeHash(p, sr);
      while (fn !== 0 && (fn & 1) === 0) {
        fn >>>= 1;
        sn >>>= 1;
      }
    } else {
      sr = nodeHash(sr, p);
    }
    fn >>>= 1;
    sn >>>= 1;
  }

  if (sn !== 0) return false;
  return equalBytes(fr, firstRoot) && equalBytes(sr, secondRoot);
}

File: sdk/python/src/proof_wire/merkle.py
"""RFC 6962 Merkle trees, proofs and their verification.

A line-for-line port of packages/core/src/merkle.js, and checked against the
same Certificate Transparency reference tree in the tests.
"""

from __future__ import annotations

from typing import Sequence

from .hashing import LEAF_PREFIX, NODE_PREFIX, equal_bytes, sha256

EMPTY_ROOT = sha256(b"")


def leaf_hash(data: bytes) -> bytes:
    return sha256(LEAF_PREFIX, data)


def node_hash(left: bytes, right: bytes) -> bytes:
    return sha256(NODE_PREFIX, left, right)


def _split_point(n: int) -> int:
    """The largest power of two strictly less than n."""
    k = 1
    while k << 1 < n:
        k <<= 1
    return k


def merkle_root(leaves: Sequence[bytes]) -> bytes:
    if not leaves:
        return EMPTY_ROOT
    if len(leaves) == 1:
        return leaves[0]
    k = _split_point(len(leaves))
    return node_hash(merkle_root(leaves[:k]), merkle_root(leaves[k:]))


def inclusion_proof(leaves: Sequence[bytes], index: int) -> list[bytes]:
    if not isinstance(index, int) or index < 0 or index >= len(leaves):
        raise IndexError(f"leaf index {index} out of range for {len(leaves)} leaves")
    if len(leaves) == 1:
        return []
    k = _split_point(len(leaves))
    if index < k:
        return inclusion_proof(leaves[:k], index) + [merkle_root(leaves[k:])]
    return inclusion_proof(leaves[k:], index - k) + [merkle_root(leaves[:k])]


def consistency_proof(leaves: Sequence[bytes], from_size: int) -> list[bytes]:
    n = len(leaves)
    if not isinstance(from_size, int) or from_size < 0 or from_size > n:
        raise IndexError(f"from_size {from_size} out of range for {n} leaves")
    if from_size in (0, n):
        return []
    return _sub_proof(leaves, from_size, True)


def _sub_proof(leaves: Sequence[bytes], m: int, is_root_of_old: bool) -> list[bytes]:
    n = len(leaves)
    if m == n:
        return [] if is_root_of_old else [merkle_root(leaves)]
    k = _split_point(n)
    if m <= k:
        return _sub_proof(leaves[:k], m, is_root_of_old) + [merkle_root(leaves[k:])]
    return _sub_proof(leaves[k:], m - k, False) + [merkle_root(leaves[:k])]


def verify_inclusion(leaf: bytes, index: int, tree_size: int, proof: Sequence[bytes], root: bytes) -> bool:
    if not isinstance(index, int) or not isinstance(tree_size, int):
        return False
    if index < 0 or tree_size <= 0 or index >= tree_size:
        return False
    fn, sn, acc = index, tree_size - 1, leaf
    for sibling in proof:
        if sn == 0 or len(sibling) != 32:
            return False
        if fn & 1 or fn == sn:
            acc = node_hash(sibling, acc)
            while fn and not fn & 1:
                fn >>= 1
                sn >>= 1
        else:
            acc = node_hash(acc, sibling)
        fn >>= 1
        sn >>= 1
    return sn == 0 and equal_bytes(acc, root)


def verify_consistency(
    first_size: int, second_size: int, first_root: bytes, second_root: bytes, proof: Sequence[bytes]
) -> bool:
    if not isinstance(first_size, int) or not isinstance(second_size, int):
        return False
    if first_size < 0 or second_size < first_size:
        return False
    if first_size == second_size:
        return len(proof) == 0 and equal_bytes(first_root, second_root)
    if first_size == 0:
        return len(proof) == 0
    fn, sn = first_size - 1, second_size - 1
    while fn & 1:
        fn >>= 1
        sn >>= 1
    if not proof:
        return False
    i = 0
    if fn:
        fr = sr = proof[0]
        i = 1
    else:
        fr = sr = first_root
    for p in proof[i:]:
        if sn == 0 or len(p) != 32:
            return False
        if fn & 1 or fn == sn:
            fr = node_hash(p, fr)
            sr = node_hash(p, sr)
            while fn and not fn & 1:
                fn >>= 1
                sn >>= 1
        else:
            sr = node_hash(sr, p)
        fn >>= 1
        sn >>= 1
    return sn == 0 and equal_bytes(fr, first_root) and equal_bytes(sr, second_root)


class MerkleTree:
    """Append-only tree with an O(log n) incremental root."""

    def __init__(self, leaves: Sequence[bytes] = ()) -> None:
        self.leaves: list[bytes] = []
        self._stack: list[tuple[int, bytes]] = []
        self._levels: list[list[bytes]] | None = None
        for leaf in leaves:
            self.append(leaf)

    def _build_levels(self) -> list[list[bytes]]:
        """Every level of the full tree, bottom up; an odd node rises unpaired.

        Makes a full-size inclusion proof O(log n) rather than O(n), which is
        the difference between exporting a large log in seconds or in hours.
        """
        if self._levels is None:
            levels = [list(self.leaves)]
            while len(levels[-1]) > 1:
                below = levels[-1]
                levels.append(
                    [node_hash(below[i], below[i + 1]) if i + 1 < len(below) else below[i] for i in range(0, len(below), 2)]
                )
            self._levels = levels
        return self._levels

    @property
    def size(self) -> int:
        return len(self.leaves)

    def append(self, h: bytes) -> int:
        self.leaves.append(h)
        self._levels = None
        self._stack.append((1, h))
        while len(self._stack) > 1 and self._stack[-1][0] == self._stack[-2][0]:
            (ls, lh), (rs, rh) = self._stack[-2], self._stack[-1]
            self._stack[-2:] = [(ls + rs, node_hash(lh, rh))]
        return len(self.leaves) - 1

    @property
    def root(self) -> bytes:
        if not self._stack:
            return EMPTY_ROOT
        acc = self._stack[-1][1]
        for _, h in reversed(self._stack[:-1]):
            acc = node_hash(h, acc)
        return acc

    def root_at(self, size: int) -> bytes:
        if size < 0 or size > len(self.leaves):
            raise IndexError(f"size {size} out of range for tree of {len(self.leaves)}")
        return merkle_root(self.leaves[:size])

    def inclusion_proof(self, index: int, tree_size: int | None = None) -> list[bytes]:
        if tree_size is not None and tree_size != len(self.leaves):
            return inclusion_proof(self.leaves[:tree_size], index)
        if not isinstance(index, int) or index < 0 or index >= len(self.leaves):
            raise IndexError(f"leaf index {index} out of range for {len(self.leaves)} leaves")
        levels = self._build_levels()
        proof, i = [], index
        for level in levels[:-1]:
            sibling = i ^ 1
            if sibling < len(level):
                proof.append(level[sibling])
            i >>= 1
        return proof

    def consistency_proof(self, from_size: int, to_size: int | None = None) -> list[bytes]:
        n = len(self.leaves) if to_size is None else to_size
        return consistency_proof(self.leaves[:n], from_size)
